File: 0347-top-k-frequent-elements/0347-top-k-frequent-elements.py
# Time: O(n) -> iterate through entire input array to insert into map
# Space: O(k) -> map of k length
class Solution:
    def topKFrequent(self, nums: List[int], k: int) -> List[int]:
        map = {}
        buckets = [[] for i in range(len(nums) + 1)] # make buckets for length of input array
        
        for n in nums:
            map[n] = 1 + map.get(n, 0) # add count to map
        for n, c in map.items():
            buckets[c].append(n) # n appears c number of times
            
        res = [] # output
        for i in range(len(buckets) - 1, 0, -1): # iterate backwards
            # get the top 'k' elements
            for n in buckets[i]:
                res.append(n)
                if len(res) == k:
                    return res

File: 0347-top-k-frequent-elements/0347-top-k-frequent-elements.js
/**
 * @param {number[]} nums
 * @param {number} k
 * @return {number[]}
 * Time: O(n) -> iterate through entire input array to insert into map
 * Space: O(k) -> map of k length
 */
const topKFrequent = (nums, k) => {
    const map = {}; // element: count
    
    for (let num of nums) {
        if (!map[num]) { // haven't seen number before
            map[num] = 0; // insert into map
        }
        map[num]++; // increment
    }
    
    // make buckets for length of input array
    const bucket = [];
    for (let i = 0; i < nums.length + 1; i++) {
        bucket.push([]);
    }
    
    // fill up buckets
    for (let key in map) {
        bucket[map[key]].push(key);
    }
    
    let res = [];
    for (let i = bucket.length - 1; i >= 0; i--) { // iterate backwards
        if (bucket[i].length > 0) { // not empty
            res = [...res, ...bucket[i]]; // get everything in res and in bucket
            if (res.length === k){
                return res;
            }
        }
    }
};